File: c++/day11/part2.cpp
#include "part2.h"

using namespace std;

/**
 * When we do not divide the worry level by 3 at every inspection, the worry level gets very big quickly.
 * This worry level is only used to decide what monkey an item is thrown to, which is decided by the 
 * remainder of the division by the test value.
 * This means we do not need to keep the original worry level, we can use any value that would produce
 * the same remainder for that dision.
 * The easiest option is to divide the worry level by the product of all test values and keep only the remainder.
 */

namespace Day11 {
    namespace Part2 {

        long long solve(const vector<Monkey>& monkeys) {
            vector<Monkey> monkeysCopy { monkeys };
            vector<long long> inspected(monkeys.size(), 0LL);

            int multiplier = 1;
            for_each(monkeys.begin(), monkeys.end(), [&](const Monkey& monkey) {
                multiplier *= monkey.test[0];
            });

            for (int i = 0; i < 10'000; ++i) {
                Part1::playRound(monkeysCopy, inspected, multiplier);
            }
            sort(inspected.begin(), inspected.end(), greater<long long>());
            return inspected[0] * inspected[1];
        }

    }
}

File: python/day11/part2.py
#!/usr/bin/env python3
import time
import part1

# without the "divide by 3" operation, we get unmanageably big worry levels
# we do not need to work with these actual numbers, since all we care about to
# decide which monkey will receive each item is whether or not the number is divisible
# by each monkey's test value
# instead of using the actual number, we can just use its remainder modulo LCM
# (least common multiple of all monkeys' test values)
# since all monkeys test values are prime, the LCM is the product of distinct values


def solve_part2(monkeys):
    lcm = 1
    for prime in set([m.dest[0] for m in monkeys]):
        lcm *= prime

    for i in range(10000):
        for monkey in monkeys:
            while len(monkey.items):
                monkey.inspected += 1
                item = monkey.items[0]
                monkey.items = monkey.items[1:]
                val = item if monkey.op[1] == 'old' else int(monkey.op[1])
                worry = item + val if monkey.op[0] == 'ADD' else item * val
                worry = worry % lcm
                next_monkey = monkey.dest[1] if worry % monkey.dest[0] == 0 else monkey.dest[2]
                monkeys[next_monkey].items.append(worry)
    inspected = sorted([monkey.inspected for monkey in monkeys], reverse=True)
    return inspected[0] * inspected[1]


if __name__ == "__main__":
    start = time.time()
    parsed = part1.parse("data.txt")
    result = solve_part2(parsed)
    print(f'Part 2 : {result}  ({round(1000 * (time.time() - start))} ms)')
